src/assets/ManifestAsset.js

```diff
@@ -0,0 +1,168 @@
+const path = require('path');
+const Asset = require('../Asset');
+
+/**
+ * A shared asset that handles:
+ * - PWA .webmanifest
+ * - PWA manifest.json
+ * - WebExtension manifest.json
+ */
+class ManifestAsset extends Asset {
+  constructor(name, options) {
+    super(name, options);
+    this.type =
+      path.basename(name) === 'manifest.json' ? 'json' : 'webmanifest';
+    this.isAstDirty = false;
+  }
+
+  parse(code) {
+    return JSON.parse(code);
+  }
+
+  processSingleDependency(path, opts) {
+    opts = opts || {entry: true};
+    return this.addURLDependency(path, opts);
+  }
+
+  processMultipleDependencies(filenames, opts) {
+    return filenames.map(filename =>
+      this.processSingleDependency(filename, opts)
+    );
+  }
+
+  processBackground(nodeName) {
+    if (nodeName !== 'background') {
+      return;
+    }
+
+    const background = this.ast[nodeName];
+    if (Array.isArray(background.scripts)) {
+      background.scripts = this.processMultipleDependencies(background.scripts);
+      this.isAstDirty = true;
+    }
+    if (background.page) {
+      background.page = this.processSingleDependency(background.page);
+      this.isAstDirty = true;
```

Comment:
You might consider pulling `this.isAstDirty = true` into `processSingleDependency` to reduce duplication. All the other `process*` methods funnel through `processSingleDependency`, so it seems like a good place to set the dirty check.